File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/random/crockford/CrockfordDammRandomStringGenerator.java
/*
 *
 *  *  Copyright (c) 2020 NHSX
 *
 */
package uk.nhs.nhsx.core.random.crockford;

import com.google.common.io.BaseEncoding;

import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.SecureRandom;
import java.util.List;
import java.util.regex.Pattern;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toList;

public class CrockfordDammRandomStringGenerator {

    private static final String RAW_BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
    private static final String CROCKFORD_BASE32_ALPHABET = "0123456789abcdefghjkmnpqrstvwxyz";

    private final DammChecksum checksum = checksum();
    private final BaseEncoding BASE_ENCODER = BaseEncoding.base32();
    private final SecureRandom random;
    private final List<Pattern> disallowedRegexes;

    public CrockfordDammRandomStringGenerator() {
        this(readBannedWordsFromClasspath());
    }

    public CrockfordDammRandomStringGenerator(List<String> bannedWordsRegexList) {
        this(new SecureRandom(), bannedWordsRegexList);
    }

    public CrockfordDammRandomStringGenerator(SecureRandom random, List<String> bannedWordsRegexList) {
        this.random = random;
        this.disallowedRegexes = bannedWordsRegexList.stream()
            .map(Pattern::compile)
            .collect(toList());
    }

    public String generate() {
        while (true) {
            byte[] bytes = new byte[5]; // ceil((7 * 5) / 8)
            random.nextBytes(bytes);
            String rawBase32 = BASE_ENCODER.encode(bytes).substring(0, 7);
            String friendlyCode = convertCrockford(rawBase32);

            String linkingId = friendlyCode + checksum.checksum(friendlyCode);
            if (!isBannedWord(linkingId)) {
                return linkingId;
            }
        }
    }

    private static List<String> readBannedWordsFromClasspath() {
        try {
            URL resource = CrockfordDammRandomStringGenerator.class.getResource("banned-words-regex-list.txt");
            return Files.readAllLines(Paths.get(requireNonNull(resource).toURI()), UTF_8);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static DammChecksum checksum() {
        return new DammChecksum(CROCKFORD_BASE32_ALPHABET);
    }

    public static class DammChecksum {

        private static final int DAMM_MODULUS = 32;
        private static final int DAMM_MASK = 5;

        private final String alphabet;

        public DammChecksum(String alphabet) {
            this.alphabet = alphabet;
        }

        public boolean validate(String string) {
            try {
                if (string == null || string.length() == 0 || string.length() > 10) return false;
                return checksum(string) == '0';
            } catch (StringIndexOutOfBoundsException e) {
                // character given was not in allowed character set
                return false;
            }
        }

        // See https://stackoverflow.com/questions/23431621/extending-the-damm-algorithm-to-base-32
        private char checksum(String linkingId) {
            return alphabet.charAt(linkingId.chars()
                .map(alphabet::indexOf)
                .reduce(0, this::dammOperation));
        }

        private int dammOperation(int checksum, int digit) {
            checksum ^= digit;
            checksum <<= 1;
            if (checksum >= DAMM_MODULUS) {
                checksum = (checksum ^ DAMM_MASK) % DAMM_MODULUS;
            }
            return checksum;
        }
    }

    private String convertCrockford(String rawBase32) {
        StringBuilder builder = new StringBuilder(rawBase32);
        for (int i = 0; i < rawBase32.length(); i++) {
            char raw = rawBase32.charAt(i);
            int value = RAW_BASE32_ALPHABET.indexOf(raw);
            builder.setCharAt(i, CROCKFORD_BASE32_ALPHABET.charAt(value));
        }
        return builder.toString();
    }

    public boolean isBannedWord(String linkingId) {
        String normalisedLinkingId = linkingId
            .replace('4', 'a')
            .replace('8', 'b')
            .replace('3', 'e')
            .replace('9', 'g')
            .replace('5', 's');
        return disallowedRegexes.stream().anyMatch(p -> p.matcher(normalisedLinkingId).find());
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/random/crockford/CrockfordDammRandomStringGenerator.kt
package uk.nhs.nhsx.core.random.crockford

import org.apache.commons.codec.binary.Base32
import java.security.SecureRandom
import java.util.regex.Pattern

class CrockfordDammRandomStringGenerator(
    private val random: SecureRandom = SecureRandom(),
    bannedWordsRegexList: List<String> = readBannedWordsFromClasspath()
) {
    private val checksum = checksum()
    private val baseEncoder = Base32(0)
    private val disallowedRegexes: List<Pattern> = bannedWordsRegexList.map { Pattern.compile(it) }

    fun generate(): String {
        while (true) {
            val bytes = ByteArray(5) // ceil((7 * 5) / 8)
            random.nextBytes(bytes)
            val rawBase32 = baseEncoder.encodeAsString(bytes).substring(0, 7)
            val friendlyCode = convertCrockford(rawBase32)
            val linkingId = friendlyCode + checksum.checksum(friendlyCode)
            if (!isBannedWord(linkingId)) {
                return linkingId
            }
        }
    }

    private fun convertCrockford(rawBase32: String): String {
        val builder = StringBuilder(rawBase32)
        for (i in rawBase32.indices) {
            val raw = rawBase32[i]
            val value = RAW_BASE32_ALPHABET.indexOf(raw)
            builder.setCharAt(i, CROCKFORD_BASE32_ALPHABET[value])
        }
        return builder.toString()
    }

    fun isBannedWord(linkingId: String): Boolean {
        val normalisedLinkingId = linkingId
            .replace('4', 'a')
            .replace('8', 'b')
            .replace('3', 'e')
            .replace('9', 'g')
            .replace('5', 's')
        return disallowedRegexes.any { it.matcher(normalisedLinkingId).find() }
    }

    companion object {
        private const val RAW_BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        private const val CROCKFORD_BASE32_ALPHABET = "0123456789abcdefghjkmnpqrstvwxyz"

        fun checksum(): DammChecksum = DammChecksum(CROCKFORD_BASE32_ALPHABET)

        private fun readBannedWordsFromClasspath(): List<String> =
            CrockfordDammRandomStringGenerator::class.java
                .getResourceAsStream("banned-words-regex-list.txt")!!
                .reader()
                .readLines()
    }
}

class DammChecksum(private val alphabet: String) {
    fun validate(string: String): Boolean = try {
        when {
            string.isEmpty() || string.length > 10 -> false
            else -> checksum(string) == '0'
        }
    } catch (e: StringIndexOutOfBoundsException) {
        // character given was not in allowed character set
        false
    }

    // See https://stackoverflow.com/questions/23431621/extending-the-damm-algorithm-to-base-32
    fun checksum(linkingId: String): Char = alphabet[linkingId.chars()
        .map { alphabet.indexOf(it.toChar()) }
        .reduce(0) { checksum, digit -> dammOperation(checksum, digit) }]

    private fun dammOperation(checksum: Int, digit: Int): Int {
        var result = checksum
        result = result xor digit
        result = result shl 1
        if (result >= DAMM_MODULUS) {
            result = (result xor DAMM_MASK) % DAMM_MODULUS
        }
        return result
    }

    companion object {
        private const val DAMM_MODULUS = 32
        private const val DAMM_MASK = 5
    }
}
